### PR TITLE
feat: add postrun hook

### PR DESCRIPTION
**Pull request by colbymatthiassonpublic**
added post run hook capability

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -355,6 +355,7 @@ export class Config implements IConfig {
     const command = c.load()
     await this.runHook('prerun', {Command: command, argv})
     await command.run(argv, this)
+    await this.runHook('postrun', {Command: command, argv})
```
Comment (amphro) [inline src/config.ts:358]: We should capture the results of the command and pass that into the port run hook. I would imagine a lot of post hooks would want to try and do something with the command results.

File: test/typescript.test.ts [hunk 2]
```diff
@@ -17,14 +17,12 @@ describe('typescript', () => {
       commandsDir: p('src/commands'),
     })
   })
-
   withConfig
   .stdout()
-  .it('runs ts command and prerun hooks', async ctx => {
+  .it('runs ts command and prerun and postrun hooks', async ctx => {
```
Comment (amphro) [inline test/typescript.test.ts:22]: Might as well have this as a separate test. Also, what about if the command errors? Add a test for that.  I'm assuming the hook won't fire. Is that the right behavior?

File: test/typescript.test.ts [hunk 3]
```diff
@@ -17,14 +17,28 @@ describe('typescript', () => {
...
+    } catch {
+      console.log('caught error')
+    }
+    expect(ctx.stdout).to.equal('running ts prerun hook\nit fails!\ncaught error\n')
+  })
+  withConfig
+  .stdout()
+  .it('runs ts command, postrun hook captures command result', async ctx => {
+    await ctx.config.runCommand('foo:bar:test-result')
+    expect(ctx.stdout).to.equal('running ts prerun hook\nit works!\nrunning ts postrun hook\nreturned success!\n')
```
Comment (amphro) [inline test/typescript.test.ts:40]: 👍

File: package.json [hunk 4]
```diff
@@ -50,6 +50,7 @@
     "posttest": "yarn lint",
     "prepack": "yarn run build",
     "test": "mocha --forbid-only \"test/**/*.test.ts\"",
+    "hooktest": "mocha --forbid-only \"test/typescript.test.ts\"",
```
Comment (amphro) [inline package.json:53]: I don't love this. The `typescript.test.ts` has a test other than hooks too. And if you need to run one file, just run mocha directly. Let's remove this.

File: src/hooks.ts [hunk 5]
```diff
@@ -10,6 +10,11 @@ export interface Hooks {
     Command: Config.Command.Class;
     argv: string[];
   };
+  postrun: {
+    Command: Config.Command.Class;
+    result: any | undefined;
```
Comment (RasPhilCo) [inline src/hooks.ts:15]:
```suggestion
    result?: any
```

File: test/typescript.test.ts [hunk 6]
```diff
@@ -17,14 +17,28 @@ describe('typescript', () => {
       commandsDir: p('src/commands'),
     })
   })
-
   withConfig
```
Comment (RasPhilCo) [inline test/typescript.test.ts:20]: add newline above

File: test/typescript.test.ts [hunk 7]
```diff
@@ -17,14 +17,28 @@ describe('typescript', () => {
...
   .stdout()
-  .it('runs ts command and prerun hooks', async ctx => {
+  .it('runs ts command, prerun and postrun hooks both trigger', async ctx => {
     await ctx.config.runCommand('foo:bar:baz')
-    expect(ctx.stdout).to.equal('running ts prerun hook\nit works!\n')
+    expect(ctx.stdout).to.equal('running ts prerun hook\nit works!\nrunning ts postrun hook\n')
+  })
+  withConfig
```
Comment (RasPhilCo) [inline test/typescript.test.ts:26]: add newline above

File: test/typescript.test.ts [hunk 8]
```diff
@@ -17,14 +17,28 @@ describe('typescript', () => {
...
+  .stdout()
+  .it('runs faulty command, only prerun hook triggers', async ctx => {
+    try {
+      await ctx.config.runCommand('foo:bar:fail')
+    } catch {
+      console.log('caught error')
+    }
+    expect(ctx.stdout).to.equal('running ts prerun hook\nit fails!\ncaught error\n')
+  })
+  withConfig
```
Comment (RasPhilCo) [inline test/typescript.test.ts:36]: add newline above

File: test/typescript.test.ts [hunk 9]
```diff
@@ -17,14 +17,28 @@ describe('typescript', () => {
       commandsDir: p('src/commands'),
     })
   })
-
   withConfig
   .stdout()
-  .it('runs ts command and prerun hooks', async ctx => {
+  .it('runs ts command, prerun and postrun hooks both trigger', async ctx => {
```
Comment (RasPhilCo) [inline test/typescript.test.ts:22]: how about? `'runs ts command and prerun & postrun hooks'`

**Comment by RasPhilCo** [comment]
@colbymatthiassonpublic @moberhauer click the CLA link to get this status check approved

> Thanks for the contribution! It looks like @colbymatthiassonpublic @moberhauer is an internal user so signing the CLA is not required. However, [we need to confirm this](https://docs.google.com/forms/d/e/1FAIpQLSfB_vIrwFLweXIGt7ID98JhISCU2zm2ErOugAnGAcCvAzr7NQ/viewform).

**Pull request merged**